R/type-data-frame.R

```diff
@@ -15,9 +15,18 @@
 #' @keywords internal
 #' @examples
 #' new_data_frame(list(x = 1:10, y = 10:1))
-new_data_frame <- function(x = list(), n = NULL, ..., class = character()) {
-  .Call(vctrs_new_data_frame, x, n, pairlist(...), class)
+new_data_frame <- function(x = list(), n = NULL, ..., class = character(0)) {
+  .Call(vctrs_new_data_frame, x, n, list(...), class)
 }
+new_data_frame <- fn_inline_formals(new_data_frame, c("x", "n", "class"))
+
+#' @rdname new_data_frame
+#' @export
+new_bare_data_frame <- function(x = list(), n = NULL) {
```

Comment:
We probably can use `if (missing(...))` to fix the speed of `new_data_frame()`.

Reply:
@lionel- and I were surprised to see that `missing(...)` is about as slow as `list(...)`, so it ends up being equivalent or slightly worse even, especially in the case when you do have attributes.

Decided to leave it as is

``` r
library(vctrs)

fn1 <- function(...) {missing(...)}
fn2 <- function(...) {list(...)}
fn3 <- function(...) {NULL}

bench::mark(fn1(), fn2(), fn3(), iterations = 100000, check = FALSE)
#> # A tibble: 3 x 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 fn1()         160ns    208ns  2426393.        0B     48.5
#> 2 fn2()         177ns    228ns  3060499.        0B     30.6
#> 3 fn3()         116ns    148ns  5603786.        0B      0
```

<sup>Created on 2020-02-18 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>